go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -364,3 +632,497 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+//		return err
+//	}
+//	if _, err := task.lifecycle.after(nextState).notify(ctx, args); err != nil {
+//		return err
+//	}
+//
+// Is roughly equivalent to:
+//
+//	if _, err := task.Lifecycle.tryAdvance(ctx, args, nextState, fn); err != nil {
+//	  return err
+//	}
+func (vtl *vcopierCopyTaskLifecycle) tryAdvance(
+	ctx context.Context,
+	args *vcopierCopyTaskArgs,
+	nextState vcopierCopyTaskState,
+	fn func(context.Context, *vcopierCopyTaskArgs) error,
+) (vcopierCopyTaskState, error) {
+	var err error
+	newState := nextState
+
+	if err = vtl.before(nextState).notify(ctx, args); err != nil {
+		goto END
+	}
+	if err = fn(ctx, args); err != nil {
+		goto END
+	}
+	if err = vtl.after(nextState).notify(ctx, args); err != nil {
+		goto END
+	}
+
+END:
+	if err != nil {
+		newState = vcopierCopyTaskFail
+		if errors.Is(err, context.Canceled) ||
+			errors.Is(err, context.DeadlineExceeded) {
+			newState = vcopierCopyTaskCancel
+		}
+	}
+
+	return newState, err
+}
+
+// do registers a callback with the vcopierCopyTaskResultHooks, to be triggered
+// when a task advances to a "done" state (= canceled, completed, failed).
+func (vrh *vcopierCopyTaskResultHooks) do(fn func(context.Context, *vcopierCopyTaskResult)) {
+	vrh.fns = append(vrh.fns, fn)
+}
+
+// notify triggers all callbacks registered with this vcopierCopyTaskResultHooks.
+func (vrh *vcopierCopyTaskResultHooks) notify(ctx context.Context, result *vcopierCopyTaskResult) {
+	for _, fn := range vrh.fns {
+		fn(ctx, result)
+	}
+}
+
+// sendTo registers a hook that accepts a result and sends the result to the
+// provided channel. E.g.:
+//
+//	resultCh := make(chan *vcopierCopyTaskResult, 1)
+//	task.lifecycle.onResult().sendTo(resultCh)
+//	defer func() {
+//	  result := <-resultCh
+//	}()
+func (vrh *vcopierCopyTaskResultHooks) sendTo(ch chan<- *vcopierCopyTaskResult) {
+	vrh.do(func(ctx context.Context, result *vcopierCopyTaskResult) {
+		select {
+		case ch <- result:
+		case <-ctx.Done():
+			// Log a warning?
+		}
+	})
+}
+
+// awaitCompletion registers a callback that returns an error unless the
+// provided chan produces a vcopierTaskResult in a complete state.
+//
+// This is useful for sequencing vcopierCopyTasks, e.g.:
+//
+//	resultCh := make(chan *vcopierCopyTaskResult, 1)
+//	prevT.lifecycle.onResult().sendTo(resultCh)
+//	currT.Lifecycle.before(vcopierCopyTaskUpdateCopyState).awaitCompletion(resultCh)
+func (vth *vcopierCopyTaskHooks) awaitCompletion(resultCh <-chan *vcopierCopyTaskResult) {
+	vth.do(func(ctx context.Context, args *vcopierCopyTaskArgs) error {
+		select {
+		case result := <-resultCh:
+			if result == nil {
+				return fmt.Errorf("channel was closed before a result received")
+			}
+			if !result.state.isDone() {
+				return fmt.Errorf("received result is not done")
+			}
+			if result.state != vcopierCopyTaskComplete {
+				return fmt.Errorf("received result is not complete")
+			}
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
```

Comment:
Matt: "this cancelation could happen from a PRS, which closes engines when a tablet transitions. We'll typically call `cancel()` on various contexts. In this case, how does this cancelation get propagated back to the parent loop.`

Reply:
Max: "this `awaitCompletion` hook is triggered in `tryAdvance`. This `ctx.Err()` will be captured by `tryAdvance()`, which will return that `err` along with `vcopierCopyTaskCancel` to `vcopierCopyWorker.execute()`, which will publish a failure result to any channels that are subscribed to task results. The "global" `resultCh` is one such subscriber, which will see that the task was canceled and return `io.EOF` while inside the `VStreamRows` loop, or `nil` when outside the `VStreamRows` loop."